tools/pythonpkg/src/include/duckdb_python/vector_conversion.hpp

```diff
@@ -47,6 +47,8 @@ struct PandasColumnBindData {
 	unique_ptr<NumPyArrayWrapper> mask;
 	// Only for categorical types
 	string internal_categorical_type;
+	// When object types are cast we must hold their data somewhere
+	vector<unique_ptr<PythonObjectContainer<py::str>>> object_str_val;
```

Comment:
Grabbing and releasing the GIL once per tuple is not ideal, could we expand the `PythonObjectContainer` to hold a vector of `py::str` objects? We could then grab the GIL once, and destroy all the objects at once.

The assign could be cleaned up by having a  `PythonGILWrapper` class that grabs the GIL in the constructor and releases it in the destructor. The PythonObjectContainer could then be extended with the following methods:


```cpp
unique_ptr<PythonGILWrapper> GetLock();
void PushInternal(PythonGILWrapper &lock, py::obj obj);
void Push(py::obj obj) {
   auto lock = GetLock();
   PushInternal(*lock, move(obj));
}
```

The `NumpyToDuckDB` can then do something like this to avoid constant acquisition/release of the GIL:

```cpp
unique_ptr<PythonGILWrapper> gil;
for(idx_t row = 0; row < count; row++) {
   ...
   if (convert_to_string) {
      if (!gil) {
         gil = obj.GetLock();
      }
      PushInternal(*gil, py::str(obj));     
   }
   ...
}
```

That way the GIL is only acquired at most once per vector, rather than once per value.